File: 0901-1000/0992-Subarrays with K Different Integers/0992-Subarrays with K Different Integers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& A, int K) {
        return atMostK(A, K) - atMostK(A, K - 1);
    }
    
private:
    int atMostK(vector<int>& A, int k) {
        vector<int> table(1 + A.size());
        int num = 0, count = 0, start = 0;
        for (int i = 0; i < A.size(); ++i) {
            if (table[A[i]]++ == 0) {
                ++num;
            }
            while (num > k) {
                if (--table[A[start++]] == 0) {
                    --num;
                }
            }
            count += i - start + 1;
        }
        return count;
    }
};

File: 0901-1000/0992-Subarrays with K Different Integers/0992-Subarrays with K Different Integers.py
import collections
class Solution:
    def subarraysWithKDistinct(self, A: List[int], K: int) -> int:
        def atMostK(k):
            table = collections.Counter()
            num = count = start = 0
            for i, a in enumerate(A):
                table[a] += 1
                if table[a] == 1:
                    num += 1
                while num > k:
                    table[A[start]] -= 1
                    if table[A[start]] == 0:
                        num -= 1
                    start += 1
                count += i - start + 1
            return count
        
        return atMostK(K) - atMostK(K - 1)
